crash when importing resource metal_virtual_circuit

```
❯ terraform import metal_virtual_circuit.connection_attachment uuid
metal_virtual_circuit.connection_attachment: Importing from ID "uuid"...
metal_virtual_circuit.connection_attachment: Import prepared!
  Prepared metal_virtual_circuit for import
metal_virtual_circuit.connection_attachment: Refreshing state... [id=uuid]
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.
╵

 


Stack trace from the terraform-provider-metal_v2.1.0 plugin:

 

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x18d03b8]

 

goroutine 81 [running]:
github.com/equinix/terraform-provider-metal/metal.resourceMetalVirtualCircuitRead(0xc0003284d0, 0x19d8080, 0xc000528500, 0xc0003284d0, 0x0)
        github.com/equinix/terraform-provider-metal/metal/resource_metal_virtual_circuit.go:134 +0x2d8
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00050fb00, 0xc000092b90, 0x19d8080, 0xc000528500, 0xc000326120, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk@v1.0.0/helper/schema/resource.go:455 +0x12e
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc000508100, 0x1c0e290, 0xc0003fcf90, 0xc0000929b0, 0xc000508100, 0xc0003fcf90, 0xc0005c3a50)
        github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/helper/plugin/grpc_provider.go:525 +0x3dd
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler(0x1a74460, 0xc000508100, 0x1c0e290, 0xc0003fcf90, 0xc000500840, 0x0, 0x1c0e290, 0xc0003fcf90, 0xc000314180, 0xb3)
        github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/tfplugin5/tfplugin5.pb.go:3153 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000cc000, 0x1c16478, 0xc000602480, 0xc000310000, 0xc00009d620, 0x20e3e70, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.23.0/server.go:995 +0x482
google.golang.org/grpc.(*Server).handleStream(0xc0000cc000, 0x1c16478, 0xc000602480, 0xc000310000, 0x0)
        google.golang.org/grpc@v1.23.0/server.go:1275 +0xd2c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003fa000, 0xc0000cc000, 0x1c16478, 0xc000602480, 0xc000310000)
        google.golang.org/grpc@v1.23.0/server.go:710 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.23.0/server.go:708 +0xa5

Error: The terraform-provider-metal_v2.1.0 plugin crashed!
```

_Reported by @christianhxc_

--- a/docs/data-sources/connection.md
+++ b/docs/data-sources/connection.md
@@ -31,6 +31,7 @@ data "metal_connection" "example" {
 * `status` - Status of the connection
 * `token` - Fabric Token for the [Equinix Fabric Portal](https://ecxfabric.equinix.com/dashboard)
 * `type` - Connection type, dedicated or shared
+* `mode` - Mode for connections in IBX facilities with the dedicated type - standard or tunnel
 * `redundancy` - Connection redundancy, reduntant or primary
 * `speed` - Connection speed in bits per second
 * `ports` - List of connection ports - primary (`ports[0]`) and secondary (`ports[1]`)

--- a/docs/data-sources/connection.md
+++ b/docs/data-sources/connection.md
@@ -29,6 +29,7 @@ data "metal_connection" "example" {
 * `facility` - Slug of a facility to which the connection belongs
 * `metro` - Slug of a metro to which the connection belongs
 * `organization_id` - ID of organization to which the connection belongs
+* `project_id` - ID of project to which the connection belongs
 * `status` - Status of the connection
 * `token` - Fabric Token for the [Equinix Fabric Portal](https://ecxfabric.equinix.com/dashboard)
 * `type` - Connection type, dedicated or shared

--- a/docs/data-sources/connection.md
+++ b/docs/data-sources/connection.md
@@ -25,6 +25,7 @@ data "metal_connection" "example" {
 
 * `description` - Description of the connection resource
 * `name` - Name of the connection resource
+* `tags` - String list of tags
 * `facility` - Slug of a facility to which the connection belongs
 * `metro` - Slug of a metro to which the connection belongs
 * `organization_id` - ID of organization to which the connection belongs

--- a/docs/data-sources/virtual_circuit.md
+++ b/docs/data-sources/virtual_circuit.md
@@ -25,11 +25,13 @@ data "metal_virtual_circuit" "example_vc" {
 ## Argument Reference
 
 * `virtual_circuit_id` - (Required) ID of the virtual circuit resource
-
 ## Attributes Reference
 
 * `name` - Name of the virtual circuit resource
 * `status` - Status of the virtal circuit
 * `project_id` - ID of project to which the VC belongs
 * `vnid`, `nni_vlan`, `nni_nvid` - VLAN parameters, see the [documentation for Equinix Fabric](https://metal.equinix.com/developers/docs/networking/fabric/)
+* `description` - Description for the Virtual Circuit resource
+* `tags` - Tags for the Virtual Circuit resource
+* `speed` - Speed of the Virtual Circuit resource
 

--- a/docs/resources/connection.md
+++ b/docs/resources/connection.md
@@ -28,6 +28,7 @@ resource "metal_connection" "test" {
 * `organization_id` - (Required) ID of the organization responsible for the connection
 * `redundancy` - (Required) Connection redundancy - redundant or primary
 * `type` - (Required) Connection type - dedicated or shared
+* `mode` - Mode for connections in IBX facilities with the dedicated type - standard or tunnel
 * `project_id` - (Optional) ID of the project where the connection is scoped to, must be set for shared connection
 * `metro` - (Optional) Metro where the connection will be created
 * `facility` - (Optional) Facility where the connection will be created

--- a/docs/resources/connection.md
+++ b/docs/resources/connection.md
@@ -33,6 +33,7 @@ resource "metal_connection" "test" {
 * `metro` - (Optional) Metro where the connection will be created
 * `facility` - (Optional) Facility where the connection will be created
 * `description` - (Optional) Description for the connection resource
+* `tags` - (Optional) String list of tags
 
 ## Attributes Reference
 

--- a/docs/resources/virtual_circuit.md
+++ b/docs/resources/virtual_circuit.md
@@ -46,7 +46,9 @@ resource "metal_virtual_circuit" "test" {
 * `vlan_id` - (Required) UUID of the VLAN to associate
 * `name` - (Optional) Name of the Virtual Circuit resource
 * `facility` - (Optional) Facility where the connection will be created
-* `description` - (Optional) Description for the connection resource
+* `description` - (Optional) Description for the Virtual Circuit resource
+* `tags` - (Optional) Tags for the Virtual Circuit resource
+* `speed` - (Optional) Speed of the Virtual Circuit resource
 
 ## Attributes Reference
 

--- a/metal/datasource_metal_connection.go
+++ b/metal/datasource_metal_connection.go
@@ -105,6 +105,11 @@ func dataSourceMetalConnection() *schema.Resource {
 				Computed:    true,
 				Description: "Connection type - dedicated or shared",
 			},
+			"mode": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Mode for connections in IBX facilities with the dedicated type - standard or tunnel",
+			},
 			"speed": {
 				Type:        schema.TypeInt,
 				Computed:    true,
@@ -168,6 +173,8 @@ func dataSourceMetalConnectionRead(d *schema.ResourceData, meta interface{}) err
 	d.Set("token", conn.Token)
 	d.Set("type", conn.Type)
 	d.Set("speed", conn.Speed)
+	d.Set("mode", conn.Mode)
+
 	d.Set("ports", getConnectionPorts(conn.Ports))
 	d.SetId(conn.ID)
 

--- a/metal/datasource_metal_connection.go
+++ b/metal/datasource_metal_connection.go
@@ -70,6 +70,12 @@ func dataSourceMetalConnection() *schema.Resource {
 				Computed:    true,
 				Description: "Description of the connection resource",
 			},
+			"tags": {
+				Type:        schema.TypeList,
+				Description: "Tags attached to the connection",
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
 			"organization_id": {
 				Type:        schema.TypeString,
 				Computed:    true,
@@ -162,6 +168,7 @@ func dataSourceMetalConnectionRead(d *schema.ResourceData, meta interface{}) err
 	d.Set("organization_id", conn.Organization.ID)
 	d.Set("name", conn.Name)
 	d.Set("description", conn.Description)
+	d.Set("tags", conn.Tags)
 	d.Set("status", conn.Status)
 	d.Set("redundancy", conn.Redundancy)
 	if conn.Facility != nil {

--- a/metal/datasource_metal_connection.go
+++ b/metal/datasource_metal_connection.go
@@ -1,8 +1,6 @@
 package metal
 
 import (
-	"strings"
-
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/packethost/packngo"
 )
@@ -81,6 +79,11 @@ func dataSourceMetalConnection() *schema.Resource {
 				Computed:    true,
 				Description: "ID of organization to which the connection belongs",
 			},
+			"project_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "ID of project to which the connection belongs",
+			},
 			"status": {
 				Type:        schema.TypeString,
 				Computed:    true,
@@ -154,36 +157,7 @@ func getConnectionPorts(cps []packngo.ConnectionPort) []map[string]interface{} {
 }
 
 func dataSourceMetalConnectionRead(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*packngo.Client)
 	connId := d.Get("connection_id").(string)
-
-	conn, _, err := client.Connections.Get(
-		connId,
-		&packngo.GetOptions{Includes: []string{"organization", "facility", "metro"}})
-	if err != nil {
-		return err
-	}
-
-	d.Set("connection_id", conn.ID)
-	d.Set("organization_id", conn.Organization.ID)
-	d.Set("name", conn.Name)
-	d.Set("description", conn.Description)
-	d.Set("tags", conn.Tags)
-	d.Set("status", conn.Status)
-	d.Set("redundancy", conn.Redundancy)
-	if conn.Facility != nil {
-		d.Set("facility", conn.Facility.Code)
-	}
-	if conn.Metro != nil {
-		d.Set("metro", strings.ToLower(conn.Metro.Code))
-	}
-	d.Set("token", conn.Token)
-	d.Set("type", conn.Type)
-	d.Set("speed", conn.Speed)
-	d.Set("mode", conn.Mode)
-
-	d.Set("ports", getConnectionPorts(conn.Ports))
-	d.SetId(conn.ID)
-
-	return nil
+	d.SetId(connId)
+	return resourceMetalConnectionRead(d, meta)
 }

--- a/metal/datasource_metal_virtual_circuit.go
+++ b/metal/datasource_metal_virtual_circuit.go
@@ -20,6 +20,17 @@ func dataSourceMetalVirtualCircuit() *schema.Resource {
 				Computed:    true,
 				Description: "Name of the virtual circuit",
 			},
+			"description": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Description of the virtual circuit",
+			},
+			"tags": {
+				Type:        schema.TypeList,
+				Description: "Tags attached to the virtual circuit",
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
 			"status": {
 				Type:        schema.TypeString,
 				Computed:    true,
@@ -67,6 +78,13 @@ func dataSourceMetalVirtualCircuitRead(d *schema.ResourceData, meta interface{})
 	d.Set("nni_vnid", vc.NniVNID)
 	d.Set("nni_vlan", vc.NniVLAN)
 	d.Set("project_id", vc.Project.ID)
+	/** TODO: packngo changes
+	d.Set("description", vc.Description)
+	tags := d.Get("tags.#").(int)
+	if tags > 0 {
+		vncr.Tags = convertStringArr(d.Get("tags").([]interface{}))
+	}
+	**/
 	d.SetId(vc.ID)
 
 	return nil

--- a/metal/resource_metal_connection.go
+++ b/metal/resource_metal_connection.go
@@ -2,6 +2,7 @@ package metal
 
 import (
 	"fmt"
+	"reflect"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/packethost/packngo"
@@ -12,6 +13,7 @@ func resourceMetalConnection() *schema.Resource {
 		Read:   resourceMetalConnectionRead,
 		Create: resourceMetalConnectionCreate,
 		Delete: resourceMetalConnectionDelete,
+		Update: resourceMetalConnectionUpdate,
 		Importer: &schema.ResourceImporter{
 			State: schema.ImportStatePassthrough,
 		},
@@ -52,6 +54,10 @@ func resourceMetalConnection() *schema.Resource {
 				Required:    true,
 				Description: "Connection type - dedicated or shared",
 				ForceNew:    true,
+			},
+			"mode": {
+				Type:        schema.TypeString,
+				Description: "Mode for connections in IBX facilities with the dedicated type - standard or tunnel",
 			},
 			"organization_id": {
 				Type:        schema.TypeString,
@@ -109,6 +115,7 @@ func resourceMetalConnectionCreate(d *schema.ResourceData, meta interface{}) err
 
 	project, projectOk := d.GetOk("project_id")
 	connType := packngo.ConnectionType(d.Get("type").(string))
+	connMode := packngo.ConnectionMode(d.Get("mode").(string))
 
 	if connType == packngo.ConnectionShared && !projectOk {
 		return fmt.Errorf("When you create a \"shared\" connection, you must set project_id")
@@ -126,6 +133,8 @@ func resourceMetalConnectionCreate(d *schema.ResourceData, meta interface{}) err
 	if connType == packngo.ConnectionShared {
 		connReq.Project = project.(string)
 	}
+
+	connReq.Mode = connMode
 
 	if metOk {
 		connReq.Metro = metro.(string)
@@ -149,6 +158,61 @@ func resourceMetalConnectionCreate(d *schema.ResourceData, meta interface{}) err
 
 	d.SetId(conn.ID)
 
+	return resourceMetalConnectionRead(d, meta)
+}
+
+func resourceMetalConnectionUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+
+	if d.HasChange("locked") {
+		var action func(string) (*packngo.Response, error)
+		if d.Get("locked").(bool) {
+			action = client.Devices.Lock
+		} else {
+			action = client.Devices.Unlock
+		}
+		if _, err := action(d.Id()); err != nil {
+			return friendlyError(err)
+		}
+	}
+	ur := packngo.ConnectionUpdateRequest{}
+
+	if d.HasChange("description") {
+		desc := d.Get("description").(string)
+		ur.Description = &desc
+	}
+
+	if d.HasChange("mode") {
+		mode := packngo.ConnectionMode(d.Get("mode").(string))
+		ur.Mode = &mode
+	}
+
+	if d.HasChange("redundancy") {
+		redundancy := packngo.ConnectionRedundancy(d.Get("redundancy").(string))
+		ur.Redundancy = redundancy
+	}
+
+	if d.HasChange("tags") {
+		ts := d.Get("tags")
+		sts := []string{}
+
+		switch ts.(type) {
+		case []interface{}:
+			for _, v := range ts.([]interface{}) {
+				sts = append(sts, v.(string))
+			}
+			ur.Tags = sts
+		default:
+			return friendlyError(fmt.Errorf("garbage in tags: %s", ts))
+		}
+	}
+
+	if !reflect.DeepEqual(ur, packngo.ConnectionUpdateRequest{}) {
+		if _, _, err := client.Connections.Update(d.Id(), &ur, nil); err != nil {
+			return friendlyError(err)
+		}
+
+	}
 	return resourceMetalConnectionRead(d, meta)
 }
 
@@ -183,6 +247,7 @@ func resourceMetalConnectionRead(d *schema.ResourceData, meta interface{}) error
 		"type":            conn.Type,
 		"speed":           conn.Speed,
 		"ports":           getConnectionPorts(conn.Ports),
+		"mode":            conn.Mode,
 	})
 }
 

--- a/metal/resource_metal_connection.go
+++ b/metal/resource_metal_connection.go
@@ -58,6 +58,8 @@ func resourceMetalConnection() *schema.Resource {
 			"mode": {
 				Type:        schema.TypeString,
 				Description: "Mode for connections in IBX facilities with the dedicated type - standard or tunnel",
+				Optional:    true,
+				Default:     "standard",
 			},
 			"organization_id": {
 				Type:        schema.TypeString,
@@ -75,8 +77,6 @@ func resourceMetalConnection() *schema.Resource {
 				Type:        schema.TypeString,
 				Optional:    true,
 				Description: "Description of the connection resource",
-				// TODO: remove ForceNew and do Update, https://github.com/packethost/packngo/issues/270
-				ForceNew: true,
 			},
 			"status": {
 				Type:        schema.TypeString,

--- a/metal/resource_metal_connection.go
+++ b/metal/resource_metal_connection.go
@@ -43,11 +43,9 @@ func resourceMetalConnection() *schema.Resource {
 				StateFunc:     toLower,
 			},
 			"redundancy": {
-				// TODO: remove ForceNew and do Update, https://github.com/packethost/packngo/issues/270
 				Type:        schema.TypeString,
 				Required:    true,
 				Description: "Connection redundancy - redundant or primary",
-				ForceNew:    true,
 			},
 			"type": {
 				Type:        schema.TypeString,
@@ -98,6 +96,12 @@ func resourceMetalConnection() *schema.Resource {
 				Elem:        connectionPortSchema(),
 				Computed:    true,
 				Description: "List of connection ports - primary (`ports[0]`) and secondary (`ports[1]`)",
+			},
+			"tags": {
+				Type:        schema.TypeList,
+				Description: "Tags attached to the connection",
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
 			},
 		},
 	}

--- a/metal/resource_metal_connection.go
+++ b/metal/resource_metal_connection.go
@@ -5,6 +5,7 @@ import (
 	"reflect"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 	"github.com/packethost/packngo"
 )
 
@@ -46,18 +47,27 @@ func resourceMetalConnection() *schema.Resource {
 				Type:        schema.TypeString,
 				Required:    true,
 				Description: "Connection redundancy - redundant or primary",
+				ValidateFunc: validation.StringInSlice([]string{
+					string(packngo.ConnectionRedundant),
+					string(packngo.ConnectionPrimary)}, false),
 			},
 			"type": {
 				Type:        schema.TypeString,
 				Required:    true,
 				Description: "Connection type - dedicated or shared",
 				ForceNew:    true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(packngo.ConnectionDedicated),
+					string(packngo.ConnectionShared)}, false),
 			},
 			"mode": {
 				Type:        schema.TypeString,
 				Description: "Mode for connections in IBX facilities with the dedicated type - standard or tunnel",
 				Optional:    true,
 				Default:     "standard",
+				ValidateFunc: validation.StringInSlice([]string{
+					string(packngo.ConnectionModeStandard),
+					string(packngo.ConnectionModeTunnel)}, false),
 			},
 			"organization_id": {
 				Type:        schema.TypeString,
@@ -132,6 +142,12 @@ func resourceMetalConnectionCreate(d *schema.ResourceData, meta interface{}) err
 		Name:       d.Get("name").(string),
 		Redundancy: packngo.ConnectionRedundancy(d.Get("redundancy").(string)),
 		Type:       connType,
+	}
+
+	// this could be generalized, see $ grep "d.Get(\"tags" *
+	tags := d.Get("tags.#").(int)
+	if tags > 0 {
+		connReq.Tags = convertStringArr(d.Get("tags").([]interface{}))
 	}
 
 	if connType == packngo.ConnectionShared {
@@ -238,6 +254,11 @@ func resourceMetalConnectionRead(d *schema.ResourceData, meta interface{}) error
 		projectId = conn.Ports[0].VirtualCircuits[0].Project.ID
 	}
 
+	mode := "standard"
+	if conn.Mode != nil {
+		mode = string(*conn.Mode)
+	}
+
 	return setMap(d, map[string]interface{}{
 		"organization_id": conn.Organization.ID,
 		"project_id":      projectId,
@@ -251,7 +272,8 @@ func resourceMetalConnectionRead(d *schema.ResourceData, meta interface{}) error
 		"type":            conn.Type,
 		"speed":           conn.Speed,
 		"ports":           getConnectionPorts(conn.Ports),
-		"mode":            conn.Mode,
+		"mode":            mode,
+		"tags":            conn.Tags,
 	})
 }
 

--- a/metal/resource_metal_connection_test.go
+++ b/metal/resource_metal_connection_test.go
@@ -106,7 +106,7 @@ func TestAccMetalConnection_Dedicated(t *testing.T) {
 				Config: testAccMetalConnectionConfig_Dedicated(rs),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("metal_connection.test", "metro", "sv"),
-					resource.TestCheckResourceAttr("metal_connection.test", "tags.#", "2"),
+					resource.TestCheckResourceAttr("metal_connection.test", "tags.#", "1"),
 					resource.TestCheckResourceAttr("metal_connection.test", "mode", "standard"),
 					resource.TestCheckResourceAttr("metal_connection.test", "type", "dedicated"),
 					resource.TestCheckResourceAttr("metal_connection.test", "redundancy", "redundant"),
@@ -122,12 +122,54 @@ func TestAccMetalConnection_Dedicated(t *testing.T) {
 				Config: testAccMetalConnectionConfig_Dedicated(rs) + testDataSourceMetalConnectionConfig_Dedicated(),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("data.metal_connection.test", "metro", "sv"),
-					resource.TestCheckResourceAttr("data.etal_connection.test", "tags.#", "2"),
+					resource.TestCheckResourceAttr("data.metal_connection.test", "tags.#", "1"),
 					resource.TestCheckResourceAttr("data.metal_connection.test", "mode", "standard"),
 					resource.TestCheckResourceAttr("data.metal_connection.test", "type", "dedicated"),
 					resource.TestCheckResourceAttr("data.metal_connection.test", "redundancy", "redundant"),
 					resource.TestCheckResourceAttr("data.metal_connection.test", "metro", "sv"),
 				),
+			},
+		},
+	})
+}
+
+func testAccMetalConnectionConfig_Tunnel(randstr string) string {
+	return fmt.Sprintf(`
+        resource "metal_project" "test" {
+            name = "tfacc-conn-pro-%s"
+        }
+
+        resource "metal_connection" "test" {
+            name            = "tfacc-conn-%s"
+            organization_id = metal_project.test.organization_id
+            metro           = "sv"
+            redundancy      = "redundant"
+            type            = "dedicated"
+            mode            = "tunnel"
+        }`,
+		randstr, randstr)
+}
+
+func TestAccMetalConnection_Tunnel(t *testing.T) {
+
+	rs := acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckMetalConnectionDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccMetalConnectionConfig_Tunnel(rs),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"metal_connection.test", "mode", "tunnel"),
+				),
+			},
+			{
+				ResourceName:      "metal_connection.test",
+				ImportState:       true,
+				ImportStateVerify: true,
 			},
 		},
 	})

--- a/metal/resource_metal_connection_test.go
+++ b/metal/resource_metal_connection_test.go
@@ -80,8 +80,17 @@ func testAccMetalConnectionConfig_Dedicated(randstr string) string {
             metro           = "sv"
             redundancy      = "redundant"
             type            = "dedicated"
+			tags            = ["tfacc"]
+			mode            = "standard"
         }`,
 		randstr, randstr)
+}
+
+func testDataSourceMetalConnectionConfig_Dedicated() string {
+	return `
+		data "metal_connection" "test" {
+            connection_id = metal_connection.test.id
+        }`
 }
 
 func TestAccMetalConnection_Dedicated(t *testing.T) {
@@ -96,14 +105,29 @@ func TestAccMetalConnection_Dedicated(t *testing.T) {
 			{
 				Config: testAccMetalConnectionConfig_Dedicated(rs),
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr(
-						"metal_connection.test", "metro", "sv"),
+					resource.TestCheckResourceAttr("metal_connection.test", "metro", "sv"),
+					resource.TestCheckResourceAttr("metal_connection.test", "tags.#", "2"),
+					resource.TestCheckResourceAttr("metal_connection.test", "mode", "standard"),
+					resource.TestCheckResourceAttr("metal_connection.test", "type", "dedicated"),
+					resource.TestCheckResourceAttr("metal_connection.test", "redundancy", "redundant"),
+					resource.TestCheckResourceAttr("metal_connection.test", "metro", "sv"),
 				),
 			},
 			{
 				ResourceName:      "metal_connection.test",
 				ImportState:       true,
 				ImportStateVerify: true,
+			},
+			{
+				Config: testAccMetalConnectionConfig_Dedicated(rs) + testDataSourceMetalConnectionConfig_Dedicated(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.metal_connection.test", "metro", "sv"),
+					resource.TestCheckResourceAttr("data.etal_connection.test", "tags.#", "2"),
+					resource.TestCheckResourceAttr("data.metal_connection.test", "mode", "standard"),
+					resource.TestCheckResourceAttr("data.metal_connection.test", "type", "dedicated"),
+					resource.TestCheckResourceAttr("data.metal_connection.test", "redundancy", "redundant"),
+					resource.TestCheckResourceAttr("data.metal_connection.test", "metro", "sv"),
+				),
 			},
 		},
 	})

--- a/metal/resource_metal_virtual_circuit.go
+++ b/metal/resource_metal_virtual_circuit.go
@@ -131,12 +131,17 @@ func resourceMetalVirtualCircuitRead(d *schema.ResourceData, meta interface{}) e
 		//"connection_id": vc.Connection.ID,
 		"project_id": vc.Project.ID,
 		"port_id":    vc.Port.ID,
-		"vlan_id":    vc.VirtualNetwork.ID,
-		"status":     vc.Status,
-		"nni_vlan":   vc.NniVLAN,
-		"vnid":       vc.VNID,
-		"nni_vnid":   vc.NniVNID,
-		"name":       vc.Name,
+		"vlan_id": func(d *schema.ResourceData, k string) error {
+			if vc.VirtualNetwork != nil {
+				return d.Set(k, vc.VirtualNetwork.ID)
+			}
+			return nil
+		},
+		"status":   vc.Status,
+		"nni_vlan": vc.NniVLAN,
+		"vnid":     vc.VNID,
+		"nni_vnid": vc.NniVNID,
+		"name":     vc.Name,
 	})
 }
 

--- a/metal/resource_metal_virtual_circuit.go
+++ b/metal/resource_metal_virtual_circuit.go
@@ -45,8 +45,8 @@ func resourceMetalVirtualCircuit() *schema.Resource {
 			},
 			"nni_vlan": {
 				Type:        schema.TypeInt,
-				Required:    true,
-				Description: "Equinix Metal network-to-network VLAN ID",
+				Description: "Equinix Metal network-to-network VLAN ID (optional when the connection has mode=tunnel)",
+				Optional:    true,
 				ForceNew:    true,
 			},
 			"vlan_id": {
@@ -78,12 +78,16 @@ func resourceMetalVirtualCircuitCreate(d *schema.ResourceData, meta interface{})
 	client := meta.(*packngo.Client)
 	vncr := packngo.VCCreateRequest{
 		VirtualNetworkID: d.Get("vlan_id").(string),
-		NniVLAN:          d.Get("nni_vlan").(int),
 		Name:             d.Get("name").(string),
 	}
+
 	connId := d.Get("connection_id").(string)
 	portId := d.Get("port_id").(string)
 	projectId := d.Get("project_id").(string)
+
+	if nniVlan, ok := d.GetOk("nni_vlan"); ok {
+		vncr.NniVLAN = nniVlan.(int)
+	}
 
 	conn, _, err := client.Connections.Get(connId, nil)
 	if err != nil {

--- a/metal/resource_metal_virtual_circuit.go
+++ b/metal/resource_metal_virtual_circuit.go
@@ -2,6 +2,7 @@ package metal
 
 import (
 	"fmt"
+	"reflect"
 	"time"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
@@ -13,6 +14,7 @@ func resourceMetalVirtualCircuit() *schema.Resource {
 	return &schema.Resource{
 		Read:   resourceMetalVirtualCircuitRead,
 		Create: resourceMetalVirtualCircuitCreate,
+		Update: resourceMetalVirtualCircuitUpdate,
 		Delete: resourceMetalVirtualCircuitDelete,
 		Importer: &schema.ResourceImporter{
 			State: schema.ImportStatePassthrough,
@@ -41,7 +43,19 @@ func resourceMetalVirtualCircuit() *schema.Resource {
 				Type:        schema.TypeString,
 				Optional:    true,
 				Description: "Name of the Virtual Circuit resource",
-				ForceNew:    true,
+				ForceNew:    true, // TODO: updateable with packngo changes
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of the Virtual Circuit resource",
+				ForceNew:    true, // TODO: updateable with packngo changes
+			},
+			"tags": {
+				Type:        schema.TypeList,
+				Description: "Tags attached to the virtual circuit",
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
 			},
 			"nni_vlan": {
 				Type:        schema.TypeInt,
@@ -79,11 +93,17 @@ func resourceMetalVirtualCircuitCreate(d *schema.ResourceData, meta interface{})
 	vncr := packngo.VCCreateRequest{
 		VirtualNetworkID: d.Get("vlan_id").(string),
 		Name:             d.Get("name").(string),
+		// Description:      d.Get("description").(string), // TODO: packngo changes
 	}
 
 	connId := d.Get("connection_id").(string)
 	portId := d.Get("port_id").(string)
 	projectId := d.Get("project_id").(string)
+
+	tags := d.Get("tags.#").(int)
+	if tags > 0 {
+		vncr.Tags = convertStringArr(d.Get("tags").([]interface{}))
+	}
 
 	if nniVlan, ok := d.GetOk("nni_vlan"); ok {
 		vncr.NniVLAN = nniVlan.(int)
@@ -146,6 +166,8 @@ func resourceMetalVirtualCircuitRead(d *schema.ResourceData, meta interface{}) e
 		"vnid":     vc.VNID,
 		"nni_vnid": vc.NniVNID,
 		"name":     vc.Name,
+		// "description": vc.Description, TODO: packngo changes
+		"tags": vc.Tags,
 	})
 }
 
@@ -168,6 +190,56 @@ func getVCStateWaiter(client *packngo.Client, id string, timeout time.Duration, 
 		MinTimeout: 5 * time.Second,
 	}
 
+}
+
+func resourceMetalVirtualCircuitUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+
+	ur := packngo.VCUpdateRequest{}
+	if d.HasChange("vnid") {
+		vnid := d.Get("vnid").(string)
+		ur.VirtualNetworkID = &vnid
+	}
+	/**
+	TODO: implement these in packngo
+	if d.HasChange("name") {
+		name := d.Get("name").(string)
+		ur.Name = &name
+	}
+
+	if d.HasChange("description") {
+		desc := d.Get("description").(string)
+		ur.Description = &desc
+	}
+
+	if d.HasChange("speed") {
+		speed := packngo.ConnectionMode(d.Get("speed").(string))
+		ur.Speed = &speed
+	}
+
+	if d.HasChange("tags") {
+		ts := d.Get("tags")
+		sts := []string{}
+
+		switch ts.(type) {
+		case []interface{}:
+			for _, v := range ts.([]interface{}) {
+				sts = append(sts, v.(string))
+			}
+			ur.Tags = sts
+		default:
+			return friendlyError(fmt.Errorf("garbage in tags: %s", ts))
+		}
+	}
+	*/
+
+	if !reflect.DeepEqual(ur, packngo.VCUpdateRequest{}) {
+		if _, _, err := client.VirtualCircuits.Update(d.Id(), &ur, nil); err != nil {
+			return friendlyError(err)
+		}
+
+	}
+	return resourceMetalVirtualCircuitRead(d, meta)
 }
 
 func resourceMetalVirtualCircuitDelete(d *schema.ResourceData, meta interface{}) error {

